File: src/main/java/com/github/kpavlov/jreactive8583/netty/pipeline/IdleEventHandler.java
package com.github.kpavlov.jreactive8583.netty.pipeline;

import com.github.kpavlov.jreactive8583.iso.MessageFactory;
import com.solab.iso8583.IsoMessage;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundHandlerAdapter;
import io.netty.handler.timeout.IdleState;
import io.netty.handler.timeout.IdleStateEvent;

public class IdleEventHandler extends ChannelInboundHandlerAdapter {

    private final MessageFactory<IsoMessage> isoMessageFactory;

    public IdleEventHandler(final MessageFactory<IsoMessage> isoMessageFactory) {
        this.isoMessageFactory = isoMessageFactory;
    }

    @Override
    public void userEventTriggered(final ChannelHandlerContext ctx, final Object evt) {
        if (evt instanceof IdleStateEvent) {
            final var e = (IdleStateEvent) evt;
            if (e.state() == IdleState.READER_IDLE || e.state() == IdleState.ALL_IDLE) {
                final var echoMessage = createEchoMessage();
                ctx.write(echoMessage);
                ctx.flush();
            }
        }
    }

    private IsoMessage createEchoMessage() {
        return isoMessageFactory.newMessage(0x800);
    }
}

File: src/main/kotlin/com/github/kpavlov/jreactive8583/netty/pipeline/IdleEventHandler.kt
package com.github.kpavlov.jreactive8583.netty.pipeline

import com.github.kpavlov.jreactive8583.iso.MessageClass
import com.github.kpavlov.jreactive8583.iso.MessageFactory
import com.github.kpavlov.jreactive8583.iso.MessageFunction
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelInboundHandlerAdapter
import io.netty.handler.timeout.IdleState
import io.netty.handler.timeout.IdleStateEvent

/**
 * IdleEventHandler sends heartbeats (administrative messages) when channel becomes idle,
 * i.e. `IdleStateEvent` is received.
 */
public open class IdleEventHandler<T>(
    private val isoMessageFactory: MessageFactory<T>,
) : ChannelInboundHandlerAdapter() {
    public override fun userEventTriggered(
        ctx: ChannelHandlerContext,
        evt: Any,
    ) {
        if (evt is IdleStateEvent &&
            (evt.state() == IdleState.READER_IDLE || evt.state() == IdleState.ALL_IDLE)
        ) {
            val heartbeatMessage = createHeartbeatMessage()
            ctx.write(heartbeatMessage)
            ctx.flush()
        }
    }

    protected fun createHeartbeatMessage(): T =
        isoMessageFactory.newMessage(
            MessageClass.NETWORK_MANAGEMENT,
            MessageFunction.REQUEST,
        )
}
